AB#695: Documentation for graphene-prepare (#31)
edgelesssys/marblerun.sh: content/docs/getting-started/cli.md

  marblerun [command]

Available Commands:
  certificate      Retrieves the certificate of the Marblerun coordinator
  graphene-prepare Modifies a Graphene manifest for use with Marblerun
  help             Help about any command
  install          Installs marblerun on a kubernetes cluster
  manifest         Manages manifest for the Marblerun coordinator
  namespace        Manages namespaces associated with Marblerun installations
  recover          Recovers the Marblerun coordinator from a sealed state
  status           Gives information about the status of the marblerun Coordinator
  uninstall        Removes Marblerun from a kubernetes cluster
  version          Display version of this CLI and (if running) the Marblerun coordinator

Flags:
  -h, --help   help for marblerun

Use "marblerun [command] --help" for more information about a command.
```
## Command `certificate`

Get the root and/or intermediate certificates of the Marblerun coordinator.

**Flags**

> These flags apply to all sub commands of certificate

Reviewer: ```suggestion
These flags apply to all `certificate` subcommands
```

| --era-config    |         | Path to remote attestation config file in json format, if none <br> provided the newest configuration will be loaded from github |
| --help, -h      |         | help for certificate                                                                                                             |
| --insecure, -i  |         | simulation mode                                                                                                                  |
| --output, -o    |         | File to save the certificate to                                                                                                  |
{{</table>}}

* ### `root`

  Gets the root certificate of the Marblerun coordinator.

  **Usage**

  ```bash
  marblerun certificate root <IP:PORT> [flags]
  ```

* ### `intermediate`

  Gets the intermediate certificate of the Marblerun coordinator.

  **Usage**

  ```bash
  marblerun certificate intermediate <IP:PORT> [flags]
  ```

* ### `chain`

  Gets the certificate chain of the Marblerun coordinator.

  **Usage**

  ```bash
  marblerun certificate chain <IP:PORT> [flags]
  ```

## Command `graphene-prepare`

> Takes a Graphene manifest template, automatically performs changes and downloads the Marblerun premain so that your application can be used in combination with Marblerun with less effort. See [Building a service: Graphene]({{< ref "docs/tasks/build-service-graphene.md" >}}) for more information.
> 
> Please note that this only works on a best-effort basis and may not instantly work correctly. While suggestions should be made for every valid TOML Graphene configuration, changes can only performed for non-hierarchically sorted configurations. as the official Graphene examples. Plus, you need to create a Marblerun manifest in addition by yourself. The unmodified manifest is saved as a backup under the old path with an added ".bak" suffix, allowing you to try out and rollback any changes performed.

Reviewer: ```suggestion
This command helps you if you want to add Graphene-based services to your Marblerun service mesh.
It prepares your Graphene project to be used as a Marble.
Given your [Graphene manifest template](https://graphene.readthedocs.io/en/latest/manifest-syntax.html), it will suggest the required adjustments needed and adds our bootstrapping data-plane code to your Graphene image.
See [Building a service: Graphene]({{< ref "docs/tasks/build-service-graphene.md" >}}) for detailed information on Marblerun’s Graphene integration and our changes in your Graphene manifest.

Please note that this only works on a best-effort basis and may not instantly work correctly.
While suggestions should be made for every valid TOML Graphene configuration, changes can only be performed for non-hierarchically sorted configurations. as the official Graphene examples.
The unmodified manifest is saved as a backup under the old path with an added ".bak" suffix, allowing you to try out and roll back any changes performed.

Remember, you need to create a [Marblerun manifest]({{< ref "docs/tasks/define-manifest.md" >}}) in addition to the Graphene manifest. Adding Graphene packages to your manifest is straightforward and follows the same principles as any other SGX enclave.
```

This command supports two modes, **spawn** and **preload**.

* **`spawn`**

>   Replaces the original entrypoint of your application with the premain process which eventually spawns your application. Dedicates argv provisioning to Marblerun, so make sure your Marblerun manifest is designed to supply the arguments correctly, otherwise you might run into issues when Graphene handled the arguments for it before.

Reviewer: ```suggestion
  Replaces the original entrypoint of your application with the bootstrapping Marble premain process which eventually spawns your application. Dedicates argv provisioning to Marblerun. If you configured the arguments to your Graphene application through the [Graphene manifest](https://graphene.readthedocs.io/en/latest/manifest-syntax.html#command-line-arguments) before, you need to transfer those to the [Marblerun manifest]({{< ref "docs/tasks/define-manifest.md#manifestmarbles">}}).
```

  **Usage**

  ```bash
  marblerun graphene-prepare spawn <path>
  ```

  **Examples**
  ```bash
  marblerun graphene-prepare spawn nginx.manifest.template
  ```

  Output:
  ```bash
  Reading file: nginx.manifest.template

  Marblerun suggests the following changes to your Graphene manifest:
  libos.entrypoint = "file:premain-graphene"
  loader.argv0_override = "$(INSTALL_DIR)/sbin/nginx"
  loader.insecure__use_host_env = 1
  sgx.allowed_files.marblerun_uuid = "file:uuid"
  sgx.enclave_size = "1024M"
  sgx.remote_attestation = 1
  sgx.thread_num = 16
  sgx.trusted_files.marblerun_premain = "file:premain-graphene"
  Do you want to automatically apply the suggested changes [y/n]? y
  Applying changes...
  Saving original manifest as nginx.manifest.template.bak...
  Saving changes to nginx.manifest.template...
  Downloading Marblerun premain from GitHub...
  Successfully downloaded premain-graphene.

  Done! You should be good to go for Marblerun!
  ```

* **`preload`**

>   Adds the premain as a shared library loaded on launch of your application via LD_PRELOAD. Allows for a faster launch than `spawn` and delegates more features to Graphene (but restricts Marblerun's functionalities), making it the quickest way to adapt your existing application. However, features such as argv provisioning cannot be used in Marblerun anymore in this mode.

Reviewer: ```suggestion
  Adds the premain as a shared library loaded during the launch of your application via LD_PRELOAD. Allows for a faster launch than `spawn` and delegates more features to Graphene (but restricts Marblerun's functionalities), making it the quickest way to adapt your existing application. However, features such as argv provisioning cannot be used in Marblerun anymore in this mode.
```

Available Commands:

>   certificate Retrieves the certificate of the Marblerun coordinator
>   check       Check the status of Marbleruns control plane
>   graphene-prepare Modifies a Graphene manifest for use with Marblerun
>   help        Help about any command
>   install     Installs marblerun on a kubernetes cluster
>   manifest    Manages manifest for the Marblerun coordinator

Reviewer 2: The spacing in this part is off
```suggestion
  certificate      Retrieves the certificate of the Marblerun coordinator
  check            Check the status of Marbleruns control plane
  graphene-prepare Modifies a Graphene manifest for use with Marblerun
  help             Help about any command
  install          Installs marblerun on a kubernetes cluster
  manifest         Manages manifest for the Marblerun coordinator
```

Author: You're right. Fixed, thanks!

  The output is similar to the following:

  ```

>   CLI Version: v0.3.0

Reviewer: Ich denke es macht nicht so wirklich Sinn hier immer die Version Strings mit jedem release zu ändern, oder seht ihr das anders?

Reviewer 2: "The output is similar to the following:"
Der commit hash ändert sich ja auch mit jeder version, ist ja nur als Beispiel wie es aussehen könnte

Author: Würde einfach 0.3.0 da stehen lassen, ist ja nur ein Example. Außer natürlich, wir ändern die Ausgabe irgendwann.
